pkg/kubectl/cmd/set/set_probe.go

```diff
@@ -0,0 +1,452 @@
+	}
+	return utilerrors.NewAggregate(errors)
+}
+
+func (o *ProbeOptions) Run(f cmdutil.Factory) error {
+	allErrs := []error{}
+	patches := CalculatePatches(o.Infos, o.Encoder, func(info *resource.Info) ([]byte, error) {
+		transformed := false
+		info.Object = info.AsVersioned()
+		_, err := o.UpdatePodSpecForObject(info.Object, func(spec *v1.PodSpec) error {
+			containers, _ := selectContainers(spec.Containers, o.ContainerSelector)
+			if len(containers) == 0 {
+				if _, err := fmt.Fprintf(o.Err, "warning: %s/%s does not have any containers matching %q\n", info.Mapping.Resource, info.Name, o.ContainerSelector); err != nil {
+					return err
+				}
+				return nil
+			}
+			// perform updates
+			transformed = true
+			for _, container := range containers {
+				o.updateContainer(container)
+			}
+			return nil
+		})
+		if transformed && err == nil {
+			return runtime.Encode(o.Encoder, info.Object)
+		}
+		return nil, err
+	})
+
+	for _, patch := range patches {
+		info := patch.Info
+		if patch.Err != nil {
+			fmt.Fprintf(o.Err, "error: %s/%s %v\n", info.Mapping.Resource, info.Name, patch.Err)
+			continue
+		}
+
+		if string(patch.Patch) == "{}" || len(patch.Patch) == 0 {
+			if _, err := fmt.Fprintf(o.Out, "%s %q was not changed\n", info.Mapping.Resource, info.Name); err != nil {
+				return err
+			}
+			continue
+		}
+		if o.PrintObject != nil && (o.Local || o.DryRun) {
+			if err := o.PrintObject(o.Cmd, o.Local, o.Mapper, patch.Info.AsVersioned(), o.Out); err != nil {
+				return err
+			}
+			continue
+		}
+		obj, err := resource.NewHelper(info.Client, info.Mapping).Patch(info.Namespace, info.Name, types.StrategicMergePatchType, patch.Patch)
+		if err != nil {
+			allErrs = append(allErrs, fmt.Errorf("failed to patch probe update to pod template: %v\n", err))
+
+			// if no port was specified, inform that one must be provided
+			if len(o.HTTPGet) > 0 && len(o.HTTPGetAction.Port.String()) == 0 {
+				fmt.Fprintf(o.Err, "\nA port must be specified as part of a url (http://127.0.0.1:3306).\nSee 'oc set probe -h' for help and examples.\n")
+			}
+			continue
+		}
+
+		// record this change (for rollout history)
+		if o.Record || cmdutil.ContainsChangeCause(info) {
+			if patch, patchType, err := cmdutil.ChangeResourcePatch(info, o.ChangeCause); err == nil {
+				if obj, err = resource.NewHelper(info.Client, info.Mapping).Patch(info.Namespace, info.Name, patchType, patch); err != nil {
+					fmt.Fprintf(o.Err, "WARNING: changes to %s/%s can't be recorded: %v\n", info.Mapping.Resource, info.Name, err)
+				}
+			}
+		}
+
+		info.Refresh(obj, true)
+
+		if len(o.Output) > 0 && o.PrintObject != nil {
+
+			if err := o.PrintObject(o.Cmd, o.Local, o.Mapper, patch.Info.AsVersioned(), o.Out); err != nil {
+				return err
+			}
+			continue
+		}
+		f.PrintSuccess(o.Mapper, o.ShortOutput, o.Out, info.Mapping.Resource, info.Name, false, "updated")
+	}
+
+	return utilerrors.NewAggregate(allErrs)
+}
```

Comment:
This whole `Run` method is almost identical to the counter parts in `set_image.go`, `set_resources.go`, `set_env.go`. Maybe we should unify this in all places instead of copy&pasting this over and over again. This in turn leads to higher maintenance costs when something breaks. @pwittrock  @adohe @mengqiy @juanvallejo wdyt?

Reply:
maybe we can handle ```func (* v1.PodSpec) error``` in the complete function. At the same time these three commands should have the same ```Option {}```.